File: src/cass/cookies.py
"""Cookie sync — extract browser cookies and push to auth service per-service."""

from __future__ import annotations

import base64
import glob
import os
import shutil
import sqlite3
import subprocess
import tempfile
import time
from pathlib import Path

import click
import httpx

from cass.config import get_default_email, require_auth


def _open_in_firefox(url: str) -> None:
    """Open a URL in Firefox specifically (cookies are extracted from Firefox)."""
    subprocess.run(["open", "-a", "Firefox", url], capture_output=True)

# Service definitions
SERVICES = {
    "yt-mcp": {
        "credential_key": "youtube_cookies",
        "domains": [".youtube.com", ".google.com"],
        "login_url": "https://accounts.google.com/ServiceLogin?service=youtube",
        "probe_url": "https://www.youtube.com/watch?v=dQw4w9WgXcQ",
        "description": "YouTube cookies for yt-dlp",
    },
    "twitter": {
        "credential_key": "twitter_cookies",
        "cookie_names": {"auth_token": "twitter_auth_token", "ct0": "twitter_ct0"},
        "domains": [".x.com", ".twitter.com"],
        "login_url": "https://x.com/i/flow/login",
        "probe_url": "https://x.com",
        "description": "Twitter/X auth cookies",
    },
    "claude-ai": {
        "credential_key": "claude_cookies",
        "domains": [".claude.ai", "claude.ai"],
        "login_url": "https://claude.ai/login",
        "probe_url": "https://claude.ai",
        "description": "Claude.ai session cookies",
    },
}


def _find_firefox_cookies_db() -> str | None:
    """Find the default Firefox profile's cookies.sqlite."""
    pattern = os.path.expanduser(
        "~/Library/Application Support/Firefox/Profiles/*/cookies.sqlite"
    )
    paths = glob.glob(pattern)
    return paths[0] if paths else None


def _check_firefox_cookies(domains: list[str], required_names: list[str] | None = None) -> bool:
    """Fast check if Firefox has cookies for the given domains. No extraction, no yt-dlp."""
    db_path = _find_firefox_cookies_db()
    if not db_path:
        return False

    tmp = tempfile.mktemp(suffix=".sqlite")
    shutil.copy2(db_path, tmp)
    try:
        conn = sqlite3.connect(tmp)
        placeholders = ",".join("?" for _ in domains)
        now = int(time.time())

        if required_names:
            # Check specific cookie names exist and aren't expired
            name_placeholders = ",".join("?" for _ in required_names)
            row = conn.execute(
                f"SELECT COUNT(DISTINCT name) FROM moz_cookies "
                f"WHERE host IN ({placeholders}) AND name IN ({name_placeholders}) "
                f"AND (expiry = 0 OR expiry > ?)",
                [*domains, *required_names, now],
            ).fetchone()
            conn.close()
            return row[0] == len(required_names)
        else:
            # Check any non-expired cookies exist for domains
            row = conn.execute(
                f"SELECT COUNT(*) FROM moz_cookies "
                f"WHERE host IN ({placeholders}) AND (expiry = 0 OR expiry > ?)",
                [*domains, now],
            ).fetchone()
            conn.close()
            return row[0] > 0
    except Exception:
        return False
    finally:
        os.unlink(tmp)


def _extract_cookies_via_ytdlp(browser: str, probe_url: str) -> list[str]:
    """Extract all browser cookies via yt-dlp, returns lines from the Netscape cookie file."""
    ytdlp = shutil.which("yt-dlp")
    if not ytdlp:
        return []

    cookies_path = Path(tempfile.mkdtemp()) / "cookies.txt"
    try:
        subprocess.run(
            [ytdlp, "--cookies-from-browser", browser, "--cookies", str(cookies_path),
             "--flat-playlist", "--skip-download", "--no-warnings", probe_url],
            capture_output=True, text=True, timeout=60,
        )
        if not cookies_path.exists() or cookies_path.stat().st_size == 0:
            return []
        return cookies_path.read_text().splitlines()
    finally:
        cookies_path.unlink(missing_ok=True)


def _filter_cookie_lines(lines: list[str], domains: list[str]) -> list[str]:
    """Filter Netscape cookie jar lines to only those matching given domains."""
    filtered = []
    for line in lines:
        if line.startswith("#") or not line.strip():
            continue
        parts = line.split("\t")
        if len(parts) < 7:
            continue
        host = parts[0]
        if any(host == d or host.endswith(d) for d in domains):
            filtered.append(line)
    return filtered


def _lines_to_jar_b64(lines: list[str]) -> str:
    """Convert filtered cookie lines to a base64-encoded Netscape cookie jar."""
    jar = "# Netscape HTTP Cookie File\n" + "\n".join(lines) + "\n"
    return base64.b64encode(jar.encode()).decode()


def _extract_named_cookies(lines: list[str], domains: list[str],
                           cookie_names: dict[str, str]) -> dict[str, str]:
    """Extract specific named cookies from Netscape cookie jar lines.

    Args:
        cookie_names: map of {cookie_name_in_jar: credential_key}
    """
    result = {}
    for line in lines:
        if line.startswith("#") or not line.strip():
            continue
        parts = line.split("\t")
        if len(parts) < 7:
            continue
        host, name, value = parts[0], parts[5], parts[6]
        if not any(host == d or host.endswith(d) for d in domains):
            continue
        if name in cookie_names:
            result[cookie_names[name]] = value
    return result


def _validate_cookies_b64(cookies_b64: str, probe_url: str) -> tuple[bool, str]:
    """Validate base64-encoded cookie jar by attempting a yt-dlp fetch.

    Returns (ok, detail) — detail is the video title on success, or error message on failure.
    """
    ytdlp = shutil.which("yt-dlp")
    if not ytdlp:
        return False, "yt-dlp not found"

    cookies_path = Path(tempfile.mkdtemp()) / "validate_cookies.txt"
    try:
        raw = base64.b64decode(cookies_b64)
        cookies_path.write_bytes(raw)
        result = subprocess.run(
            [ytdlp, "--cookies", str(cookies_path), "--skip-download",
             "--print", "title", "--no-warnings", probe_url],
            capture_output=True, text=True, timeout=30,
        )
        if result.returncode == 0 and result.stdout.strip():
            return True, result.stdout.strip()
        return False, result.stderr.strip()[:200] or "yt-dlp failed with no output"
    except Exception as exc:
        return False, str(exc)
    finally:
        cookies_path.unlink(missing_ok=True)


def _push_credentials(service: str, credentials: dict[str, str]) -> None:
    """Push credentials to auth service."""
    base_url, headers = require_auth()
    email = get_default_email()

    if "X-Auth-Secret" in headers:
        url = f"{base_url}/credentials/{email}/{service}"
        body = {"credentials": credentials}
    else:
        url = f"{base_url}/api/extension/credentials/{service}"
        body = credentials

    resp = httpx.put(url, headers=headers, json=body, timeout=15)
    resp.raise_for_status()


@click.group()
def cookies() -> None:
    """Sync browser cookies to platform services."""


@cookies.command()
@click.argument("services", nargs=-1)
@click.option("--dry-run", is_flag=True, help="Extract cookies but don't push.")
@click.option("--no-open", is_flag=True,
              help="Don't open login pages on missing/invalid cookies. For unattended use (e.g. session-start hooks).")
def sync(services: tuple[str, ...], dry_run: bool, no_open: bool) -> None:
    """Sync browser cookies to services.

    Specify service names (yt-mcp, twitter, claude-ai) or omit for all.
    Opens login pages automatically if cookies are missing (unless --no-open).
    """
    ytdlp = shutil.which("yt-dlp")
    if not ytdlp:
        raise click.ClickException("yt-dlp is required. Install: brew install yt-dlp")

    targets = list(services) if services else list(SERVICES.keys())

    for name in targets:
        if name not in SERVICES:
            click.echo(f"Unknown service: {name} (available: {', '.join(SERVICES)})")
            continue

        svc = SERVICES[name]
        click.echo(f"\n── {name}: {svc['description']} ──")
        _sync_service(name, svc, dry_run, no_open)

    click.echo()


def _sync_service(name: str, svc: dict, dry_run: bool, no_open: bool = False) -> None:
    """Extract and sync cookies for a single service."""
    def _prompt_login(msg: str) -> None:
        click.echo(f"  {msg}")
        if no_open:
            click.echo("  Sign in to Firefox, then re-run: cass cookies sync " + name)
        else:
            _open_in_firefox(svc["login_url"])
            click.echo("  Sign in, then re-run this command.")

    click.echo("  Extracting from firefox...")
    lines = _extract_cookies_via_ytdlp("firefox", svc["probe_url"])

    if not lines:
        _prompt_login("No cookies found.")
        return

    cookie_names = svc.get("cookie_names")
    if cookie_names:
        # Named cookie mode (twitter: auth_token, ct0)
        creds = _extract_named_cookies(lines, svc["domains"], cookie_names)
        if not creds:
            _prompt_login("Cookies present but missing required keys.")
            return
    else:
        # Full cookie jar mode (yt-mcp, claude-ai)
        filtered = _filter_cookie_lines(lines, svc["domains"])
        if not filtered:
            _prompt_login(f"No cookies for {', '.join(svc['domains'])}.")
            return
        creds = {svc["credential_key"]: _lines_to_jar_b64(filtered)}

    # Validate cookie jar before pushing (yt-mcp, claude-ai — full jar mode)
    cred_key = svc.get("credential_key", "")
    if cred_key in creds and svc.get("probe_url"):
        click.echo("  Validating cookies...")
        ok, detail = _validate_cookies_b64(creds[cred_key], svc["probe_url"])
        if ok:
            click.echo(f"  Valid — {detail}")
        else:
            click.echo(f"  INVALID — {detail}", err=True)
            _prompt_login("Cookies are stale or logged out.")
            return

    if dry_run:
        keys = ", ".join(creds.keys())
        click.echo(f"  Found: {keys}")
        click.echo("  Dry run — not pushing.")
        return

    _push_credentials(name, creds)
    keys = ", ".join(creds.keys())
    click.echo(f"  Synced: {keys} ✓")


@cookies.command()
def status() -> None:
    """Check which services have valid cookies in Firefox (fast, no network)."""
    missing = []
    for name, svc in SERVICES.items():
        required = list(svc["cookie_names"].keys()) if svc.get("cookie_names") else None
        ok = _check_firefox_cookies(svc["domains"], required)
        icon = "ok" if ok else "MISSING"
        click.echo(f"  {name:12s} {icon}")
        if not ok:
            missing.append(name)

    if missing:
        click.echo(f"\nRun: cass cookies sync {' '.join(missing)}")
        raise SystemExit(1)


@cookies.command()
def test() -> None:
    """Test that yt-dlp can fetch a YouTube page with Firefox cookies."""
    ytdlp = shutil.which("yt-dlp")
    if not ytdlp:
        raise click.ClickException("yt-dlp not found in PATH")

    click.echo("Testing yt-dlp with firefox cookies...")
    result = subprocess.run(
        [ytdlp, "--cookies-from-browser", "firefox", "--skip-download",
         "--print", "title", "https://www.youtube.com/watch?v=dQw4w9WgXcQ"],
        capture_output=True, text=True, timeout=30,
    )
    if result.returncode != 0:
        click.echo(f"FAIL: {result.stderr.strip()}", err=True)
        raise SystemExit(1)
    click.echo(f"OK — got title: {result.stdout.strip()}")

File: internal/cmd/cookies.go
package cmd

import (
	"bytes"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"os/exec"
	"path/filepath"
	"sort"
	"strings"
	"time"

	"github.com/spf13/cobra"

	"github.com/Cassandras-Edge/cass/internal/auth"
	"github.com/Cassandras-Edge/cass/internal/browser"
	"github.com/Cassandras-Edge/cass/internal/config"
)

type cookieService struct {
	Name           string
	CredentialKey  string            // For "full jar" mode (single base64 blob)
	CookieNames    map[string]string // For "named" mode (jar-name → credential-name)
	Domains        []string
	LoginURL       string
	ProbeURL       string
	Description    string
	RequiredCookie []string // For skip_validate services — sanity check before push
	SkipValidate   bool
}

// cookieServices is the static catalog mirrored from cookies.py SERVICES.
var cookieServices = map[string]*cookieService{
	"yt-mcp": {
		Name:          "yt-mcp",
		CredentialKey: "youtube_cookies",
		Domains:       []string{".youtube.com", ".google.com"},
		LoginURL:      "https://accounts.google.com/ServiceLogin?service=youtube",
		ProbeURL:      "https://www.youtube.com/watch?v=dQw4w9WgXcQ",
		Description:   "YouTube cookies for yt-dlp",
	},
	"twitter": {
		Name:          "twitter",
		CredentialKey: "twitter_cookies",
		CookieNames:   map[string]string{"auth_token": "twitter_auth_token", "ct0": "twitter_ct0"},
		Domains:       []string{".x.com", ".twitter.com"},
		LoginURL:      "https://x.com/i/flow/login",
		ProbeURL:      "https://x.com",
		Description:   "Twitter/X auth cookies",
	},
	"claude-ai": {
		Name:           "claude-ai",
		CredentialKey:  "claude_cookies",
		Domains:        []string{".claude.ai", "claude.ai"},
		LoginURL:       "https://claude.ai/login",
		ProbeURL:       "https://claude.ai",
		Description:    "Claude.ai session cookies",
		RequiredCookie: []string{"sessionKey"},
		SkipValidate:   true, // yt-dlp can't fetch claude.ai (not a media site)
	},
	"perplexity-mcp": {
		Name:           "perplexity-mcp",
		CredentialKey:  "perplexity_cookies",
		Domains:        []string{".perplexity.ai", "www.perplexity.ai"},
		LoginURL:       "https://www.perplexity.ai/finance/AAPL",
		ProbeURL:       "https://www.perplexity.ai/finance/AAPL",
		Description:    "Perplexity Finance cookies (cf_clearance + session)",
		RequiredCookie: []string{"cf_clearance"},
		SkipValidate:   true,
	},
}

func cookiesCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "cookies",
		Short: "Sync browser cookies to platform services",
	}
	cmd.AddCommand(cookiesSyncCmd())
	cmd.AddCommand(cookiesStatusCmd())
	cmd.AddCommand(cookiesTestCmd())
	cmd.AddCommand(cookiesRefreshCmd())
	return cmd
}

func cookiesSyncCmd() *cobra.Command {
	var dryRun, noOpen bool
	cmd := &cobra.Command{
		Use:   "sync [services...]",
		Short: "Extract Firefox cookies for one or more services and push them to auth",
		RunE: func(_ *cobra.Command, args []string) error {
			targets := args
			if len(targets) == 0 {
				for k := range cookieServices {
					targets = append(targets, k)
				}
				sort.Strings(targets)
			}
			for _, name := range targets {
				svc, ok := cookieServices[name]
				if !ok {
					known := serviceNames()
					fmt.Printf("Unknown service: %s (available: %s)\n", name, strings.Join(known, ", "))
					continue
				}
				fmt.Printf("\n── %s: %s ──\n", svc.Name, svc.Description)
				syncOneService(svc, dryRun, noOpen)
			}
			fmt.Println()
			return nil
		},
	}
	cmd.Flags().BoolVar(&dryRun, "dry-run", false, "Extract cookies but don't push")
	cmd.Flags().BoolVar(&noOpen, "no-open", false, "Don't open login pages on missing cookies (for unattended use)")
	return cmd
}

func cookiesStatusCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "status",
		Short: "Show which services have Firefox cookies present (fast, no network)",
		RunE: func(_ *cobra.Command, _ []string) error {
			// Fast path: read Firefox's moz_cookies SQLite directly. ~5ms/service
			// vs ~1-2s/service for yt-dlp shell-out. Matches Python's check.
			missing := []string{}
			for _, name := range serviceNames() {
				svc := cookieServices[name]
				var required []string
				if svc.CookieNames != nil {
					for k := range svc.CookieNames {
						required = append(required, k)
					}
				}
				ok := browser.HasFirefoxCookies(svc.Domains, required)
				icon := "ok"
				if !ok {
					icon = "MISSING"
					missing = append(missing, name)
				}
				fmt.Printf("  %-14s %s\n", name, icon)
			}
			if len(missing) > 0 {
				fmt.Printf("\nRun: cass cookies sync %s\n", strings.Join(missing, " "))
				os.Exit(1)
			}
			return nil
		},
	}
}

func cookiesTestCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "test",
		Short: "Verify that yt-dlp can fetch a YouTube page with Firefox cookies",
		RunE: func(_ *cobra.Command, _ []string) error {
			if _, err := exec.LookPath("yt-dlp"); err != nil {
				return fmt.Errorf("yt-dlp not found in PATH")
			}
			fmt.Println("Testing yt-dlp with firefox cookies...")
			out, err := exec.Command(
				"yt-dlp", "--cookies-from-browser", "firefox", "--skip-download",
				"--print", "title", "https://www.youtube.com/watch?v=dQw4w9WgXcQ",
			).CombinedOutput()
			if err != nil {
				return fmt.Errorf("FAIL: %s", strings.TrimSpace(string(out)))
			}
			fmt.Printf("OK — got title: %s\n", strings.TrimSpace(string(out)))
			return nil
		},
	}
}

func cookiesRefreshCmd() *cobra.Command {
	var timeoutSec int
	var force bool
	cmd := &cobra.Command{
		Use:   "refresh [services...]",
		Short: "Background-refresh Cloudflare-gated cookies in Firefox, then sync",
		Long: `Currently meaningful only for services with Cloudflare bot protection
(perplexity-mcp). Snapshots cf_clearance from Firefox's SQLite, opens
the probe URL in a background tab, then polls until cf_clearance
rotates (or --timeout elapses). Matches the Python behavior.`,
		RunE: func(_ *cobra.Command, args []string) error {
			targets := args
			if len(targets) == 0 {
				targets = []string{"perplexity-mcp"}
			}
			for _, name := range targets {
				svc, ok := cookieServices[name]
				if !ok {
					fmt.Printf("Unknown service: %s\n", name)
					continue
				}
				fmt.Printf("\n── refresh %s ──\n", name)
				refreshOneService(svc, timeoutSec, force)
			}
			return nil
		},
	}
	cmd.Flags().IntVar(&timeoutSec, "timeout", 30, "Seconds to wait for Firefox to refresh Cloudflare cookies")
	cmd.Flags().BoolVar(&force, "force", false, "Open Firefox even if cookies look fresh")
	return cmd
}

func refreshOneService(svc *cookieService, timeoutSec int, force bool) {
	// cf_clearance lives on perplexity.ai's host. For other services that
	// don't ride Cloudflare bot protection, the refresh is just a sync.
	var clearanceHost string
	if svc.Name == "perplexity-mcp" {
		clearanceHost = "%perplexity.ai"
	}

	var beforeValue string
	var beforeExpiry int64
	if clearanceHost != "" {
		beforeValue, beforeExpiry = browser.ReadFirefoxCookie(clearanceHost, "cf_clearance")
		if !force && beforeValue != "" {
			ttl := beforeExpiry - time.Now().Unix()
			if ttl > 3600 {
				fmt.Printf("  cf_clearance still has %d min left — running sync.\n", ttl/60)
				syncOneService(svc, false, true)
				return
			}
		}
	}

	fmt.Printf("  Opening %s in background Firefox...\n", svc.ProbeURL)
	_ = exec.Command("open", "-a", "Firefox", "-g", svc.ProbeURL).Run()

	if clearanceHost != "" {
		deadline := time.Now().Add(time.Duration(timeoutSec) * time.Second)
		refreshed := false
		for time.Now().Before(deadline) {
			time.Sleep(2 * time.Second)
			nowVal, _ := browser.ReadFirefoxCookie(clearanceHost, "cf_clearance")
			if nowVal != "" && nowVal != beforeValue {
				refreshed = true
				break
			}
		}
		if refreshed {
			fmt.Println("  cf_clearance refreshed.")
		} else {
			fmt.Fprintf(os.Stderr, "  WARN: no cf_clearance change in %ds — syncing what we have.\n", timeoutSec)
		}
	} else {
		// No Cloudflare gate — small delay then sync.
		time.Sleep(2 * time.Second)
	}
	syncOneService(svc, false, true)
}

func syncOneService(svc *cookieService, dryRun, noOpen bool) {
	promptLogin := func(msg string) {
		fmt.Printf("  %s\n", msg)
		if noOpen {
			fmt.Printf("  Sign in to Firefox, then re-run: cass cookies sync %s\n", svc.Name)
		} else {
			_ = exec.Command("open", "-a", "Firefox", svc.LoginURL).Run()
			fmt.Println("  Sign in, then re-run this command.")
		}
	}
	fmt.Println("  Extracting from firefox...")
	cookies, err := browser.ReadFirefoxCookies(svc.Domains)
	if err != nil || len(cookies) == 0 {
		promptLogin("No cookies found.")
		return
	}

	var creds map[string]string
	if svc.CookieNames != nil {
		creds = map[string]string{}
		for _, c := range cookies {
			if mapped, ok := svc.CookieNames[c.Name]; ok {
				creds[mapped] = c.Value
			}
		}
		if len(creds) == 0 {
			promptLogin("Cookies present but missing required keys.")
			return
		}
	} else {
		lines := make([]string, 0, len(cookies))
		for _, c := range cookies {
			lines = append(lines, c.NetscapeLine())
		}
		jar := "# Netscape HTTP Cookie File\n" + strings.Join(lines, "\n") + "\n"
		creds = map[string]string{svc.CredentialKey: base64.StdEncoding.EncodeToString([]byte(jar))}
	}

	if !svc.SkipValidate && svc.CredentialKey != "" && creds[svc.CredentialKey] != "" {
		if _, err := exec.LookPath("yt-dlp"); err != nil {
			fmt.Println("  (yt-dlp not installed — skipping fetch validation)")
		} else {
			fmt.Println("  Validating cookies...")
			ok, detail := validateCookiesB64(creds[svc.CredentialKey], svc.ProbeURL)
			if ok {
				fmt.Printf("  Valid — %s\n", detail)
			} else {
				fmt.Fprintf(os.Stderr, "  INVALID — %s\n", detail)
				promptLogin("Cookies are stale or logged out.")
				return
			}
		}
	}

	if svc.SkipValidate && len(svc.RequiredCookie) > 0 {
		jarB64 := creds[svc.CredentialKey]
		raw, _ := base64.StdEncoding.DecodeString(jarB64)
		missing := []string{}
		for _, n := range svc.RequiredCookie {
			if !strings.Contains(string(raw), "\t"+n+"\t") {
				missing = append(missing, n)
			}
		}
		if len(missing) > 0 {
			fmt.Fprintf(os.Stderr, "  MISSING required cookies: %s\n", strings.Join(missing, ", "))
			promptLogin(fmt.Sprintf("Visit %s in Firefox to get fresh cookies.", svc.ProbeURL))
			return
		}
	}

	if dryRun {
		keys := make([]string, 0, len(creds))
		for k := range creds {
			keys = append(keys, k)
		}
		fmt.Printf("  Found: %s\n", strings.Join(keys, ", "))
		fmt.Println("  Dry run — not pushing.")
		return
	}
	if err := pushCredentials(svc.Name, creds); err != nil {
		fmt.Fprintf(os.Stderr, "  push failed: %v\n", err)
		return
	}
	keys := make([]string, 0, len(creds))
	for k := range creds {
		keys = append(keys, k)
	}
	fmt.Printf("  Synced: %s ✓\n", strings.Join(keys, ", "))
}

func validateCookiesB64(jarB64, probeURL string) (bool, string) {
	if _, err := exec.LookPath("yt-dlp"); err != nil {
		return false, "yt-dlp not found"
	}
	tmp, err := os.MkdirTemp("", "cass-validate-")
	if err != nil {
		return false, err.Error()
	}
	defer os.RemoveAll(tmp)
	cookiesPath := filepath.Join(tmp, "cookies.txt")
	raw, err := base64.StdEncoding.DecodeString(jarB64)
	if err != nil {
		return false, "decode jar: " + err.Error()
	}
	if err := os.WriteFile(cookiesPath, raw, 0o600); err != nil {
		return false, err.Error()
	}
	out, err := exec.Command(
		"yt-dlp", "--cookies", cookiesPath, "--skip-download",
		"--print", "title", "--no-warnings", probeURL,
	).CombinedOutput()
	if err != nil || strings.TrimSpace(string(out)) == "" {
		s := strings.TrimSpace(string(out))
		if len(s) > 200 {
			s = s[:200]
		}
		if s == "" {
			s = "yt-dlp failed with no output"
		}
		return false, s
	}
	return true, strings.TrimSpace(string(out))
}

func pushCredentials(service string, credentials map[string]string) error {
	creds, err := auth.Read()
	if err != nil {
		return fmt.Errorf("not logged in (run: cass login): %w", err)
	}
	body, _ := json.Marshal(credentials)
	url := fmt.Sprintf("%s/api/extension/credentials/%s", config.PortalURL(), service)
	req, _ := http.NewRequest("PUT", url, bytes.NewReader(body))
	req.Header.Set("CF-Access-Client-Id", creds.CFAccessClientID)
	req.Header.Set("CF-Access-Client-Secret", creds.CFAccessClientSecret)
	req.Header.Set("Authorization", "Bearer "+creds.MCPKey)
	req.Header.Set("Content-Type", "application/json")
	resp, err := (&http.Client{Timeout: 15 * time.Second}).Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		buf, _ := io.ReadAll(io.LimitReader(resp.Body, 1024))
		return fmt.Errorf("portal %s: %s", resp.Status, string(buf))
	}
	// Read-back verify — portal route returns the credentials object directly.
	verifyReq, _ := http.NewRequest("GET", url, nil)
	verifyReq.Header.Set("CF-Access-Client-Id", creds.CFAccessClientID)
	verifyReq.Header.Set("CF-Access-Client-Secret", creds.CFAccessClientSecret)
	verifyReq.Header.Set("Authorization", "Bearer "+creds.MCPKey)
	vresp, err := (&http.Client{Timeout: 15 * time.Second}).Do(verifyReq)
	if err != nil {
		return fmt.Errorf("PUT succeeded but verify failed: %w", err)
	}
	defer vresp.Body.Close()
	var stored map[string]any
	if err := json.NewDecoder(vresp.Body).Decode(&stored); err == nil {
		if inner, ok := stored["credentials"].(map[string]any); ok {
			stored = inner
		}
		var missing []string
		for k, v := range credentials {
			if got, _ := stored[k].(string); got != v {
				missing = append(missing, k)
			}
		}
		if len(missing) > 0 {
			return fmt.Errorf("PUT returned 200 but read-back is missing keys: %s", strings.Join(missing, ", "))
		}
	}
	return nil
}

func serviceNames() []string {
	names := make([]string, 0, len(cookieServices))
	for k := range cookieServices {
		names = append(names, k)
	}
	sort.Strings(names)
	return names
}
